zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -19,19 +26,286 @@ use {
+                id_note INTEGER PRIMARY KEY,
+                tx INTEGER NOT NULL,
+                output_index INTEGER NOT NULL,
+                from_account INTEGER NOT NULL,
+                address TEXT NOT NULL,
+                value INTEGER NOT NULL,
+                memo BLOB,
+                FOREIGN KEY (tx) REFERENCES transactions(id_tx),
+                FOREIGN KEY (from_account) REFERENCES accounts(account),
+                CONSTRAINT tx_output UNIQUE (tx, output_index)
+            );",
+        )?;
+        Ok(())
+    }
+
+    fn down(&self, _transaction: &Transaction) -> Result<(), SqliteClientError> {
+        // We should never down-migrate the first migration, as that can irreversibly
+        // destroy data.
+        panic!("Cannot revert the initial migration.");
+    }
+}
+
+struct WalletMigration1;
+
+migration!(
+    WalletMigration1,
+    "a2e0ed2e-8852-475e-b0a4-f154b15b9dbe",
+    ["bc4f5e57-d600-4b6c-990f-b3538f0bfce1"],
+    "Add support for receiving transparent UTXOs."
+);
+
+impl RusqliteMigration for WalletMigration1 {
+    type Error = SqliteClientError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), SqliteClientError> {
+        transaction.execute_batch(
+            "CREATE TABLE IF NOT EXISTS utxos (
+                id_utxo INTEGER PRIMARY KEY,
+                address TEXT NOT NULL,
+                prevout_txid BLOB NOT NULL,
+                prevout_idx INTEGER NOT NULL,
+                script BLOB NOT NULL,
+                value_zat INTEGER NOT NULL,
+                height INTEGER NOT NULL,
+                spent_in_tx INTEGER,
+                FOREIGN KEY (spent_in_tx) REFERENCES transactions(id_tx),
+                CONSTRAINT tx_outpoint UNIQUE (prevout_txid, prevout_idx)
+            );",
+        )?;
+        Ok(())
+    }
+
+    fn down(&self, transaction: &Transaction) -> Result<(), SqliteClientError> {
+        transaction.execute_batch("DROP TABLE utxos;")?;
+        Ok(())
+    }
+}
+
+struct WalletMigration2<P: consensus::Parameters> {
+    params: P,
+    seed: SecretVec<u8>,
+}
+
+impl<P: consensus::Parameters> Migration for WalletMigration2<P> {
+    fn id(&self) -> Uuid {
+        ::uuid::Uuid::parse_str("be57ef3b-388e-42ea-97e2-678dafcf9754").unwrap()
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        ["a2e0ed2e-8852-475e-b0a4-f154b15b9dbe"]
+            .iter()
+            .map(|uuidstr| ::uuid::Uuid::parse_str(uuidstr).unwrap())
+            .collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Add support for unified full viewing keys"
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for WalletMigration2<P> {
+    type Error = SqliteClientError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), SqliteClientError> {
+        //
+        // Update the accounts table to store ufvks rather than extfvks
+        //
+
+        transaction.execute_batch(
+            "PRAGMA foreign_keys = OFF;
```

Comment:
https://www.sqlite.org/pragma.html#pragma_foreign_keys says:
> This pragma is a no-op within a transaction; foreign key constraint enforcement may only be enabled or disabled when there is no pending [BEGIN](https://www.sqlite.org/lang_transaction.html) or [SAVEPOINT](https://www.sqlite.org/lang_savepoint.html).

Reply:
I'm not sure what to think about this; I know that I had to add this pragma in order to avoid getting a foreign key constraint violation when moving the new accounts table into place.

Reply:
Huh. Maybe the sqlite documentation is wrong? It also says:
> As of SQLite [version 3.6.19](https://www.sqlite.org/releaselog/3_6_19.html), the default setting for foreign key enforcement is OFF. However, that might change in a future release of SQLite.

(`rusqlite` is [version 0.24 used with the "bundled" feature](https://github.com/zcash/librustzcash/blob/276bcc34f58dd2b97a9685e3fbe20f583c24a341/zcash_client_sqlite/Cargo.toml#L23) which I believe [bundles](https://github.com/rusqlite/rusqlite/tree/v0.24.2/libsqlite3-sys/sqlite3) sqlite 3.33.0.)

The default being OFF isn't consistent with needing to explicitly set it to OFF to avoid a constraint violation, so I'm confused. I guess there's no harm in including this pragma, though.

Reply:
Is this database shared between the Rust and Kotlin code on Android?

For some of the shared databases, our database initialization in the Android SDK is turning on foreign key constraints.  So daira is correct that SQLite has constraints turned off by default, but our SDK is turning them on for the derived data db.